tools: add Boxstarter script

bzoz opened the pull request:
Adds a [Boxstarter](http://boxstarter.org/) script and documentation for an easy way to install Node.js building prerequisites on Windows. It will install git, python, msbuild tools and optionally VSCode. 

Ref: https://github.com/nodejs/code-and-learn/issues/71

The links will not work until this lands. A version with working links can be found here: ~~[tools/boxstarter/README.md](https://github.com/janeasystems/node/blob/bartek-boxstarter-janea/tools/boxstarter/README.md)~~
Edit: new link - [tools/bootstrap/README.md](https://github.com/janeasystems/node/blob/bartek-boxstarter-janea/tools/bootstrap/README.md).

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
build, doc, tools

bzoz commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11455/

benjamingr commented:
cc @refack @mwrock 
----

Great work, I'm hugely +1 on anything that'll make onboarding windows users easier. Boxstarter which I was unfamiliar with looks based on the built in package manager and with a permissive license.

benjamingr commented on tools/boxstarter/README.md:
````diff
@@ -0,0 +1,54 @@
+Install-BoxstarterPackage https://raw.githubusercontent.com/nodejs/node/master/tools/boxstarter/node_boxstarter -DisableReboots
+```
+
+ * For Node.js prerequisites with VS Code:  
````
Not sure why we need this?

benjamingr commented on tools/boxstarter/node_boxstarter_vscode:
```diff
@@ -0,0 +1,25 @@
+#
+
+# Git and Unix tools will be added to the PATH
+choco install git -params /GitAndUnixToolsOnPath -y
```
I'd prefer it if we didn't install the editor as well - most people being onboarded have their own preference for an editor.

benjamingr commented on tools/boxstarter/README.md:
```diff
@@ -0,0 +1,54 @@
+# Boxstarter setup for Node.js
+
+A [Boxstarter][] script can be used for an easy setup of Windows systems with all
```
maybe remove `an`.

gibfahn commented on tools/boxstarter/node_boxstarter_vscode:
```diff
@@ -0,0 +1,25 @@
+choco install visualstudio2017-workload-vctools -y
+
+# Installs Visual Studio Code
+choco install visualstudiocode -y
```
+1 to not installing an editor, especially if it means duplicating the whole script.

digitalinfinity replied:
Instead of duplicating the script, we could just check if an environment variable is set indicating whether an editor should be installed? (not advocating one way or the other but just pointing out that we likely don't need to duplicate the whole script)

gibfahn commented on tools/boxstarter/node_boxstarter:
```diff
@@ -0,0 +1,21 @@
+
+# Installs VS 2017 Build Tools
+choco install visualstudio2017buildtools -y
+choco install visualstudio2017-workload-vctools -y
```
What is the benefit of using Boxstarter here rather than just using chocolatey directly? Is it the web url? 

Looking at [their page](http://boxstarter.org/WhyBoxstarter), I'm not seeing anything obviously beneficial, and it's another thing for new starters to understand.

```console
:: Copied from chocolatey website
@"%SystemRoot%\System32\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -InputFormat None -ExecutionPolicy Bypass -Command "iex ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))" && SET "PATH=%PATH%;%ALLUSERSPROFILE%\chocolatey\bin"

choco install git -params /GitAndUnixToolsOnPath -y
choco install python2 visualstudio2017buildtools visualstudio2017-workload-vctools -y
```

refack replied:
Maybe use a more generic directory name like `tools/bootstrap/` and then iterate over recipes.

joaocgreis replied:
This is an improvement either way, but I'd go with boxstarter rather than just chocolatey. For users that want to understand this or already know about chocolatey, the script is very easy to understand and use as a guide for their own custom installation (plus, dependencies are already documented in `BUILDING.md`).

Looking at the Boxstarter page, it adds a lot of resilience to the process (reboots, disabling updates, etc), moving it much closer to the ideal one-click-no-issues setup.

refack commented:
Love it.
I would go one step further and add a shortcut through `vcbuild boxstarter`:
```cmd
powershell.exe -ExecutionPolicy Unrestricted -noprofile -Command "&{iex ((New-Object System.Net.WebClient).DownloadString('http://boxstarter.org/bootstrapper.ps1')); get-boxstarter -Force; Install-BoxstarterPackage https://raw.githubusercontent.com/nodejs/node/master/tools/boxstarter/node_boxstarter -DisableReboots}"
```

digitalinfinity commented on tools/boxstarter/README.md:
```diff
@@ -0,0 +1,54 @@
+## WebLauncher Installation
+
+To install Node.js prerequisites using [Boxstarter WebLauncher][], just open
+**one** of  the following links with Internet Explorer or Edge browser on the
```
Why does it need to be just those two browsers?

bzoz replied:
It is a "Click Once" app (https://msdn.microsoft.com/en-us/library/t71a733d.aspx). It works in IE and Edge by default, other browsers need a plugin.

digitalinfinity commented:
This is awesome, thanks @bzoz! (cc @yodurr @rachelnicole). +1 to @refack's suggestion of adding this support to vcbuild too (although I would still prefer having an explicit developer shell for Node but that can come later 😄).

bzoz commented:
Updated the PR, PTAL.

I've renamed it from `tools/boxstarter` to `tools/bootstrap`. I've dropped VSCode - I've assumed one of the use cases would be installing dev environment on a completely clean Windows install, having a link that would also install editor sounded reasonable. 

@refack  `vcbuild boxstart` would be great, but it needs to be run in a elevated terminal. I guess some more powershell can do the UAC. It could be also used to disable Windows Defender for the Node folder (as mentioned in https://github.com/nodejs/node/issues/16278).

gibfahn commented:
I think you can do this to get elevated powershell:

`@"%SystemRoot%\System32\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -InputFormat None -ExecutionPolicy Bypass -Command "..."`

(copied from the chocolatey getting started.

gibfahn commented on BUILDING.md:
```diff
@@ -202,6 +202,9 @@ Prerequisites:
   and tools which can be included in the global `PATH`.
 
+*Note:* All prerequisites can be easily installed by following
+[this bootstrapping guide](https://github.com/nodejs/node/blob/master/tools/bootstrap/README.md#Windows).
```
Maybe not move this to the top of the list and drop the `#Windows` off the end? That way it'll work when we add other platforms.

bzoz replied:
Sorry, I don't understand what you mean

gibfahn replied:
Move this line to L83, and change it to:

```markdown
All prerequisites can be easily installed by following
[this bootstrapping guide](https://github.com/nodejs/node/blob/master/tools/bootstrap/README.md).
```

bzoz commented:
Updated - moved the link, and added instructions for Linux. PTAL

richardlau commented on tools/bootstrap/README.md:
```diff
@@ -22,6 +22,20 @@ get-boxstarter -Force
 
+## Linux
+
+For building Node.js on Linux, following packets are required (note, that this
```
Should `packets` be `packages`?

bzoz replied:
Yes

gibfahn commented on tools/bootstrap/README.md:
```diff
@@ -0,0 +1,46 @@
+To bootstrap Node.js on Linux, run in terminal:
+  * OpenSUSE: `sudo zypper install git-core python gcc-c++ make`
+  * Fedora: `sudo dnf install git-core python gcc-c++ make`
+  * Ubuntu, Debian: `sudo apt-get install git-core python g++ make`
```
Does git-core work? From [the website](https://git-scm.com/download/linux) it seems like it should be `git` for all three distros.

bzoz replied:
It looks like on openSUSE [git-core](https://software.opensuse.org/package/git-core) is more lightweight than [git](https://software.opensuse.org/package/git) package. But I think I'll change this anyway.

gibfahn commented on tools/bootstrap/README.md:
```diff
@@ -0,0 +1,46 @@
+  * OpenSUSE: `sudo zypper install git-core python gcc-c++ make`
+  * Fedora: `sudo dnf install git-core python gcc-c++ make`
+  * Ubuntu, Debian: `sudo apt-get install git-core python g++ make`
+
```
For macOS I'm pretty sure it's just:

```bash
xcode-select --install # Installs git, make, and clang
```

`python2` is already installed.

digitalinfinity commented:
I gave this a spin on a new machine which happened to have VS (internal dogfood) and git already installed. There were probably some issues caused by VS already being installed because the boxstarter script ended with a strange warning about chocolatey failing but didn't actually print out what the error was, it just asked me to hit enter and then it exited- but it looks like python27 was installed correctly.

One thing that might be useful to add in the readme is how much free disk space the user should be expected to have before running the script?

bzoz commented:
@digitalinfinity I just tried that, a fresh Win2016 install on amazon takes 13.7 GB. After the script it rises to 21.6 GB

digitalinfinity commented:
Great, thanks for checking! Perhaps it would be good to document that expectation?

bzoz commented:
Updated, PTAL

bzoz commented:
Landed in https://github.com/nodejs/node/commit/61d46dd7e05021ac1e5f9add415fbaf36b17d5ff
